src/Data/Bifunctor/Linear.hs

```diff
@@ -63,8 +62,5 @@ instance SymmetricMonoidal (,) () where
   rassoc ((x,y),z) = (x,(y,z))
 
 instance SymmetricMonoidal Either Void where
-  swap (Left x) = Right x
-  swap (Right x) = Left x
-  rassoc (Left (Left x)) = Left x
-  rassoc (Left (Right x)) = Right (Left x)
-  rassoc (Right x) = Right (Right x)
+  swap = either Right Left
+  rassoc = either (either Left (Right . Left)) (Right . Right)
```

Comment:
I don't really like this `rassoc` definition. It's terse, but it's hard to read. The previous one is clearer in my eyes.

The other definitions, on the other hand, are convincing!

Reply:
Ahh, something frustrating about this is that the older one doesn't compile in ghc-8.9.0.20190613, but the new one does...

Reply:
I'm confused. What is the problem?

Reply:
If I try to compile
```haskell
instance SymmetricMonoidal Either Void where
  swap = either Right Left
  rassoc (Left (Left x)) = Left x
  rassoc (Left (Right x)) = Right (Left x)
  rassoc (Right x) = Right (Right x)
```
with ghc-8.9.0.20190613, linearity errors show up:
 ```
/Users/bmehta/Programming/haskell/tweag/linear-base/src/Data/Bifunctor/Linear.hs:67:23: error:
    • Couldn't match expected multiplicity ‘'One’ of variable ‘x’ with actual multiplicity ‘'Omega’
    • In the pattern: Right x
      In the pattern: Left (Right x)
      In an equation for ‘rassoc’:
          rassoc (Left (Right x)) = Right (Left x)
   |
67 |   rassoc (Left (Right x)) = Right (Left x)
   |                       ^

/Users/bmehta/Programming/haskell/tweag/linear-base/src/Data/Bifunctor/Linear.hs:68:17: error:
    • Couldn't match expected multiplicity ‘'One’ of variable ‘x’ with actual multiplicity ‘'Omega’
    • In the pattern: Right x
      In an equation for ‘rassoc’: rassoc (Right x) = Right (Right x)
      In the instance declaration for ‘SymmetricMonoidal Either Void’
   |
68 |   rassoc (Right x) = Right (Right x)
   |                 ^
```

Reply:
Notably, there's no error for the pattern `Left`, only for `Right`.

Reply:
What compiler are you using presently?

Reply:
For general work I'm using tweag/linear-types:1.0.0 (not entirely sure how to check the exact version past that) - for some reason this started working the other day for me so I've stuck with it.

Reply:
For that error message I used 1.0.5, with `compiler: ghc-8.9` in my `stack.yaml`.

Reply:
Ok, I've spoken with @monoidal and we know what the issue is. We will fix it.

In the meantime, `rassoc (Right x) = (Right :: _ ->. _) (Right x)` will work.

(with `-XPartialTypeSignature` the warning on partial type signature deactivated, or by filling the `_`)

Reply:
Or use the defnition in terms of `either`, but I'm not terribly fond of it.

Reply:
Thanks! No rush on the fix, I'm happy working with 1.0.0

Reply:
In the next version of the image (not released yet), `Right (Right x)` should work.